### PR TITLE
Update NextJS to v12

### LINKED ISSUE #1264: Update NextJS to v12

### PR DESCRIPTION
**Pull request by SahilKalra98**
**Closes #1264**

**This PR updates Next to version 12 using the steps here**:
https://nextjs.org/docs/upgrading

**In-Progress:**

The `helpers/apolloClient-server.ts` file needs to be refactored a bit more since there is duplicate code copied from `helpers/apolloClient.ts`.

The `apolloClient-server.ts` file needed to be created because it seems like dynamic importing broke while upgrading to version 12. [This link](https://github.com/garageScript/c0d3-app/issues/1264) talks more about the issue.

---
**Version 12 Changes**
Here is a [link](https://nextjs.org/blog/next-12#es-modules-support-and-url-imports) to all of the changes introduced into Next.js 12. Some of the big changes I saw was Next transitioning from Babel to SWC (Rust-based compiler to compile JavaScript/TypeScript) which is supposed to be a lot faster than babel and also support for ES Modules. I didn't include SWC in this PR because it could make the PR bigger so I decided we could tackle this in another PR if we decide to transition to it. A link to SWC is [here](https://swc.rs/)

**Concerns/Issues**

1. The first issue I saw was when I removed the `babel.config.js` file to use SWC, the server worked fine and the files were able to get compiled however all the tests broke. Most of the errors I saw were `cannot use import` errors or jsx experimental flag is not set. I looked at the Jest docs [here](https://jestjs.io/docs/ecmascript-modules) which we can try out in another PR to avoid making this PR too large. Also at the bottom of that link it states that `jest.mock` is not supported in a clean way and is still experimental. To look at the errors when removing the `babel.config.js` you can do the following:
   1. `rm babel.config.js`
   2. `yarn test`

2. The second issue that arose was after upgrading to version 12, the `__tests__/pages/curriculum.test.js` file was not passing. The error that came up was this `ReferenceError: setImmediate is not defined` from the `nodemailer` package in `helpers/mail.ts`. In order to solve the problem I used this [link](https://github.com/prisma/prisma/issues/8558) and the method described at the bottom of the discussion. I wasn't sure if this was a good way of solving the problem, would appreciate some feedback! In order to reproduce the error you can checkout this branch and then do the following:
    1. `yarn install`
    2. `rm jest.env.js`
    3. Remove the key `setupFilesAfterEnv` inside `jest.config.js`
    4. Then run `yarn test`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,6 +36,8 @@
     "@prisma/client": "^2.29.1",
     "@quixo3/prisma-session-store": "^3.1.2",
     "@sentry/nextjs": "^6.13.3",
+    "@swc/cli": "^0.1.55",
```
Comment (songz) [inline package.json:39]: this is packaged into nextv12

File: tsconfig.json [hunk 2]
```diff
@@ -19,13 +16,9 @@
     "resolveJsonModule": true,
     "skipLibCheck": true,
     "strict": true,
-    "target": "esnext"
+    "target": "esnext",
+    "incremental": true
```
Comment (songz) [inline tsconfig.json:20]: why did you add `incremental: true`?

File: helpers/apolloClient-server.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+function createApolloClient() {
+  return new ApolloClient({
+    ssrMode: typeof window === 'undefined',
+    link: createIsomorphLink(),
+    cache
+  })
+}
+
+export function initializeApollo(
+  initialState: NormalizedCacheObject | null = null
```
Comment (songz) [inline helpers/apolloClient-server.ts:35]: if you look at [how its used](https://github.com/garageScript/c0d3-app/pull/1289/files#diff-127feee063dd438c812f7f1ddfadfe8f2b1c8e2b6a865342a668c0e7075f67c2R220), it seems like we don't need `initialState`

File: helpers/apolloClient-server.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+) {
+  const _apolloClient = apolloClient || createApolloClient()
+
+  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
+  // gets hydrated here
+  if (initialState) {
```
Comment (songz) [inline helpers/apolloClient-server.ts:41]: comment above

File: helpers/apolloClient-server.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        ...sourceArray,
+        ...destinationArray.filter(d => sourceArray.every(s => !isEqual(d, s)))
+      ]
+    })
+
+    // Restore the cache with the merged data
+    _apolloClient.cache.restore(data)
+  }
+  // For SSG and SSR always create a new Apollo Client
+  if (typeof window === 'undefined') return _apolloClient
```
Comment (songz) [inline helpers/apolloClient-server.ts:58]: we don't need this `if` check because it's server side apolloClient so window should always be `undefined`

File: helpers/apolloClient-server.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import merge from 'deepmerge'
+import isEqual from 'lodash/isEqual'
+import { SchemaLink } from '@apollo/client/link/schema'
+import { schema } from '../graphql/schema'
+
+let apolloClient: ApolloClient<NormalizedCacheObject>
+const cache = new InMemoryCache()
+
+function createIsomorphLink() {
+  if (typeof window === 'undefined') {
```
Comment (songz) [inline helpers/apolloClient-server.ts:15]: you don't need this check

File: helpers/apolloClient-server.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    return new HttpLink({
+      uri: '/api/graphql',
+      credentials: 'same-origin'
+    })
+  }
+}
+
+function createApolloClient() {
+  return new ApolloClient({
+    ssrMode: typeof window === 'undefined',
```
Comment (songz) [inline helpers/apolloClient-server.ts:28]: this could be hardcoded

File: helpers/apolloClient.ts [hunk 8]
```diff
@@ -47,7 +47,7 @@ function createIsomorphLink() {
   }
 }
 
-function createApolloClient() {
+export function createApolloClient() {
```
Comment (songz) [inline helpers/apolloClient.ts:50]: why does this need to be exported?

**Comment by songz** [comment]
could you also update `apolloClient` accordingly? seems like `window` will never be `undefined` in `apolloClient` https://github.com/garageScript/c0d3-app/pull/1289/files#diff-f73b56ae46bdf4aff9c3bcea9558552bc858e86d51a638186f395753438bfa6eR37-R40

File: tsconfig.json (hunk 2 above)
Comment (SahilKalra98) [reply tsconfig.json:20]: When I upgraded to v12, and then did `yarn dev`, the log statements said that it configured the tsconfig.json file. The only thing added was `"incremental": true`

Comment (songz) [reply tsconfig.json:20]: ah it stores info about the complied files: https://www.typescriptlang.org/tsconfig#incremental

File: helpers/apolloClient-server.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+function createApolloClient() {
+  return new ApolloClient({
+    ssrMode: true,
+    link: new SchemaLink({ schema }),
+    cache: new InMemoryCache()
+  })
+}
+
+export function initializeApollo(): ApolloClient<NormalizedCacheObject> {
+  return createApolloClient()
```
Comment (songz) [inline helpers/apolloClient-server.ts:18]: `creatApolloClient` seems like an unnecessary function. Perhaps just `return new ApolloClient({....})`?

File: helpers/apolloClient.ts [hunk 10]
```diff
@@ -33,24 +30,18 @@ const whiteList = ['/curriculum']
...
-    return new SchemaLink({ schema })
-  } else {
-    const { HttpLink } = require('@apollo/client/link/http')
-    return new HttpLink({
-      uri: '/api/graphql',
-      credentials: 'same-origin'
-    })
-  }
+function createHttpLink() {
+  const { HttpLink } = require('@apollo/client/link/http')
```
Comment (songz) [inline helpers/apolloClient.ts:34]: seems like this doesn't need to be `require`'d dynamically. Could you try importing it

File: helpers/apolloClient.ts [hunk 11]
```diff
@@ -33,24 +30,18 @@ const whiteList = ['/curriculum']
...
+  return new HttpLink({
```
Comment (songz) [inline helpers/apolloClient.ts:35]: if the import above works then we don't need this `creatHttpLink` helper function anymore

**Review by songz — APPROVED** [review]

**Review by flacial — APPROVED** [review]

File: helpers/apolloClient.ts [hunk 12]
```diff
@@ -18,10 +18,7 @@ const whiteList = ['/curriculum']
 
 //there is no global await and this logic can't be refactored into other functions because pages/_app.tsx can't be async
 ;(async function () {
-  if (
-    typeof window !== 'undefined' &&
-    whiteList.includes(window.location.pathname)
-  ) {
+  if (whiteList.includes(window.location.pathname)) {
```
Comment (flacial) [inline helpers/apolloClient.ts:21]: I think this is why Vercel build throw errors. The code first runs server-side and at that period, `window` is undefined.

The past check of whether `window` is undefined or not has to return.

```suggestion
 if (
    typeof window !== 'undefined' &&
    whiteList.includes(window.location.pathname)
    )
```

**Review by flacial — CHANGES_REQUESTED** [review]

File: helpers/apolloClient.ts (hunk 12 above)
Comment (SahilKalra98) [reply helpers/apolloClient.ts]: Thanks!

File: helpers/apolloClient.ts (hunk 10 above)
Comment (SahilKalra98) [reply helpers/apolloClient.ts:37]: I tried importing and in the testfile, it seems that MockFunction was never called [here](https://github.com/garageScript/c0d3-app/blob/master/helpers/apolloClient.test.js#L63)

**Review by flacial — APPROVED** [review]
